### PR TITLE
[Raptor] Modify back tracking journeys

### PR DESCRIPTION
**Pull request by xlqian**

### CODE REVIEW

File: source/routing/raptor_api.cpp [hunk 1]
```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
...
+    auto order = section.get_in_st->order();
+    // because of stay-ins, we may have several vj in one section, we have to scan the stop times
+    // of all vjs
+    for (const auto* vj = section.get_in_st->vehicle_journey; vj; (vj = vj->next_vj, order = type::RankStopTime(0))) {
+        for (const auto& st :
+             boost::make_iterator_range(vj->stop_time_list.begin() + order.val, vj->stop_time_list.end())) {
+            if (st.stop_point->stop_area->uri == stop_area_uri && vj != vj_to_skip) {
+                if (clockwise) {
+                    section.get_out_st = &st;
+                    section.get_out_dt = st.departure_time;
```
Comment (pbench) [inline source/routing/raptor_api.cpp]:
```suggestion
                    section.get_out_dt = st.arrival_time;
```
you "get out" at "arrival_time"

File: source/routing/raptor_api.cpp [hunk 2]
```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
...
+                } else {
+                    section.get_in_st = &st;
+                    section.get_in_dt = st.arrival_time;
```
Comment (pbench) [inline source/routing/raptor_api.cpp]:
```suggestion
                    section.get_in_dt = st.departure_time;
```
You "get in" at "departure time"

File: source/routing/raptor_api.cpp [hunk 3]
```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
...
+    for (auto& section : journey.sections) {
+        if (update_visited_section(section, stop_area_uri, vj_to_skip, clockwise)) {
+            return {true, section_idx};
+        }
+        ++section_idx;
     }
-    auto it = journeys.begin();
-    while (it != journeys.end()) {
-        const auto& journey = *it;
+    return {false, 0};
```
Comment (pbench) [inline source/routing/raptor_api.cpp:1414]: Why do you need to iterate through all sections ?
You know that you will modify either the last section (clockwise) or the first section (!clockwise)

File: source/routing/raptor_api.cpp [hunk 4]
```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
...
-             (vj = vj->next_vj, order = type::RankStopTime(0))) {
-            for (const auto& st :
-                 boost::make_iterator_range(vj->stop_time_list.begin() + order.val, vj->stop_time_list.end())) {
-                if (st.stop_point->stop_area->uri == stop_area_uri && vj != vj_to_skip) {
-                    return true;
-                }
-                if (section.get_out_st == &st) {
-                    reach_end = true;
-                    break;
+bool update_visited_section(navitia::routing::Journey::Section& section,
```
Comment (pbench) [inline source/routing/raptor_api.cpp]:
```suggestion
// - if `clockwise` : remove everything in the section *after* the *first* occurence of `stop_area_uri`, unless this stop is served by `vj_to_skip`
// - if `! clockwise` : remove everything in the section *before* the *last* occurence of `stop_area_uri`, unless this stop is served by `vj_to_skip`
//
// returns `true` if the section has been modified
bool shorten_section(navitia::routing::Journey::Section& section
```

File: source/routing/raptor_api.cpp [hunk 5]
```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
...
-    auto it = journeys.begin();
-    while (it != journeys.end()) {
-        const auto& journey = *it;
+    return {false, 0};
+}
 
-        if (!journey.is_pt()) {
-            ++it;
-            continue;
+bool find_stop_area_of_interest_through_journey(navitia::routing::Journey& journey,
```
Comment (pbench) [inline source/routing/raptor_api.cpp]:
```suggestion
// - if `clockwise` : remove everything in the journey *after* the *first* occurence of `stop_area_uri`, unless this stop is served by `vj_to_skip`
// - if `! clockwise` : remove everything in the journey *before* the *last* occurence of `stop_area_uri`, unless this stop is served by `vj_to_skip`
//
// returns `true` if the journey has been modified
bool shorten_journey(navitia::routing::Journey& journey,
```

File: source/routing/raptor_api.cpp [hunk 6]
```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
...
+    LOG4CPLUS_WARN(logger, "Error occurred when modifying backtracking journeys");
     return false;
 }
 
-void filter_backtracking_journeys(RAPTOR::Journeys& journeys, const bool clockwise) {
-    log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
-
-    if (journeys.size() == 1) {
-        return;
+std::pair<bool, size_t> get_and_update_visited_section(navitia::routing::Journey& journey,
```
Comment (pbench) [inline source/routing/raptor_api.cpp:1408]: Maybe you can inline this function inside `find_stop_area_of_interest_through_journey` ?

Comment (xlqian) [reply source/routing/raptor_api.cpp:1408]: I deliberately put this in a function so that I don't have to use `break` to stop the loop

**Review by pbench — APPROVED** [review]

**Pull request merged**